src/box/gc.c

```diff
@@ -733,6 +806,228 @@ gc_consumer_name(struct gc_consumer *consumer)
 			  tt_uuid_str(&consumer->uuid));
 }
 
+/**
+ * Write the consumer to persistent WAL GC state. The function must be called
+ * only when schema supports persistent consumers.
+ */
+static int
+gc_consumer_sync_impl(struct gc_consumer *consumer)
+{
+	assert(gc_has_persistent_consumers());
+
+	/* Mark consumer as synchronized with persistent state. */
+	consumer->is_synced = true;
+	char tuple_buf[UUID_STR_LEN + VCLOCK_STR_LEN_MAX + GC_NAME_MAX + 30];
+	char *tuple_end = tuple_buf;
+	tuple_end = mp_encode_array(tuple_end, 3);
+	tuple_end = mp_encode_str0(tuple_end, tt_uuid_str(&consumer->uuid));
+	tuple_end = mp_encode_vclock_ignore0(tuple_end, &consumer->vclock);
+	tuple_end = mp_encode_map(tuple_end, 1);
+	tuple_end = mp_encode_str0(tuple_end, "type");
+	tuple_end = mp_encode_str0(tuple_end,
+				   gc_consumer_type_strs[consumer->type]);
+	assert((unsigned long)(tuple_end - tuple_buf) < sizeof(tuple_buf));
+
+	return box_replace(BOX_GC_CONSUMERS_ID, tuple_buf, tuple_end, NULL);
+}
+
+int
+gc_consumer_sync(struct gc_consumer *consumer)
+{
+	/*
+	 * We cannot persist consumer asynchronously for the first time
+	 * because hanging consumer can appear in the case of fail. So let's
+	 * always do a direct replace here.
+	 */
+	if (gc_has_persistent_consumers()) {
+		struct txn *txn = txn_begin();
+		if (txn == NULL)
+			return -1;
+		/*
+		 * No need to wait limbo, so let's set TXN_FORCE_ASYNC.
+		 * Moreover, the function can be called when quorum cannot be
+		 * collected, (for example, on IPROTO_SUBSCRIBE), so we must set
+		 * this flag here.
+		 */
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (gc_consumer_sync_impl(consumer) != 0) {
+			txn_abort(txn);
+			return -1;
+		}
+		if (txn_commit(txn) != 0)
+			return -1;
+	}
+	/* Unref WAL GC delay if being persisted for the first time. */
+	if (!consumer->is_persistent)
+		gc_delay_unref();
+	consumer->is_persistent = true;
+	return 0;
+}
+
+int
+gc_erase_consumer(const struct tt_uuid *uuid)
+{
+	/* Delete in-memory consumer. */
+	bool was_persistent = false;
+	struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+	if (consumer != NULL) {
+		/* Consumer won't be asynchronously advanced anymore. */
+		was_persistent = consumer->is_persistent;
+		consumer->is_persistent = false;
+	}
+
+	/* Try to delete consumer from persistent state first. */
+	if (gc_has_persistent_consumers() &&
+	    recovery_state == FINISHED_RECOVERY &&
+	    boxk(IPROTO_DELETE, BOX_GC_CONSUMERS_ID, "[%s]",
```

Comment:
I just decided to put the check here because we shouldn't write to space on recovery anyway, so it's better to check the thing in one place than in many. But I don't mind to put the check to `alter.cc`, if you don't like it here.

Reply:
Done, `gc` subsystem doesn't care about recovery state anymore.

Reply:
Great, less dependencies between the modules = better.